stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+# Algorithms and Data Structures
+
+## Algorithms overview
+### What is algorithm?
+An _**algorithm**_ is a set of instructions to solve a particular problem. Any code may be called an algorithm. Computer algorithms work via input and output. They take the input and apply each step of the algorithm to that information to generate an output. For example, a search engine is an algorithm that takes a search query as an input and searches its database for items relevant to the words in the query. It then outputs the results.
+
+![An algorithm model](img/algorithm_model.png)
+
+### Measuring algorithm performance 
+While writing a solution we need to find the most efficient algorithm, when a word "efficient" means "running the least time".
+
+In computer science, a term that describes the amount of time it takes to run an algorithm is called "_**Time Complexity**_". Time complexity is commonly estimated by counting the number of elementary operations performed by the algorithm. It is supposed that each elementary operation lasts a fixed amount of time. Thus, in order to measure "time complexity", we should take into account:
+* the amount of time taken by an algorithm;
+* the number of elementary operations performed by it.
+
+#### Big O Notation
+The time complexity is commonly expressed using "Big O notation" (Landau's symbol). It is a special notation that tells you how fast an algorithm is. It’s called "Big O notation" because you put a “big O” in front of the number of operations.
+
+![Big O notation](img/big_o_notation.png "Big O notation")
+
+As regards an origin of "Big O notation", it is used in complexity theory, computer science, and mathematics to describe the asymptotic behavior of functions, and it tells you how fast a function grows or declines (you are not supposed to know it, just for your information). Letter "O" is used because the rate of growth of a function is called its ***O***rder.
+
+#### Big O Complexity Chart
+![Big O Complexity Chart](img/big_o_complexity_chart.jpeg "Big O Complexity Chart")
+
+"Big O notation" tells us a number of operations, which an algorithm will perform.
```

Comment:
I would add some example, like if the complexity is O(n) - it is approximately, could be smth like 
![image](https://user-images.githubusercontent.com/56270371/160281699-3d48b573-00ab-4780-8cfc-e2c085de246d.png)